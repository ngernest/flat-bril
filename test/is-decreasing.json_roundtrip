{
  "functions": [
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "x"
          ],
          "dest": "tmp0",
          "funcs": [
            "is_decreasing"
          ],
          "labels": [],
          "op": "call",
          "type": "bool"
        },
        {
          "args": [
            "tmp0"
          ],
          "dest": "tmp",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "bool"
        },
        {
          "args": [
            "tmp"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "x"
          ],
          "dest": "tmp",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "tmp1",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "dest": "tmp2",
          "op": "const",
          "type": "int",
          "value": -1
        },
        {
          "args": [
            "tmp1",
            "tmp2"
          ],
          "dest": "tmp3",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "tmp3"
          ],
          "dest": "prev",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "label": "label4"
        },
        {
          "dest": "tmp7",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "tmp",
            "tmp7"
          ],
          "dest": "tmp8",
          "funcs": [],
          "labels": [],
          "op": "gt",
          "type": "bool"
        },
        {
          "args": [
            "tmp8"
          ],
          "funcs": [],
          "labels": [
            "label5",
            "label6"
          ],
          "op": "br"
        },
        {
          "label": "label5"
        },
        {
          "args": [
            "tmp"
          ],
          "dest": "tmp9",
          "funcs": [
            "last_digit"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "tmp9"
          ],
          "dest": "digit",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "digit",
            "prev"
          ],
          "dest": "tmp10",
          "funcs": [],
          "labels": [],
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "tmp10"
          ],
          "funcs": [],
          "labels": [
            "label11",
            "label12"
          ],
          "op": "br"
        },
        {
          "label": "label11"
        },
        {
          "dest": "tmp14",
          "op": "const",
          "type": "bool",
          "value": false
        },
        {
          "args": [
            "tmp14"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "label13"
          ],
          "op": "jmp"
        },
        {
          "label": "label12"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "label13"
          ],
          "op": "jmp"
        },
        {
          "label": "label13"
        },
        {
          "args": [
            "digit"
          ],
          "dest": "prev",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "tmp15",
          "op": "const",
          "type": "int",
          "value": 10
        },
        {
          "args": [
            "tmp",
            "tmp15"
          ],
          "dest": "tmp16",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "tmp16"
          ],
          "dest": "tmp",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "label4"
          ],
          "op": "jmp"
        },
        {
          "label": "label6"
        },
        {
          "dest": "tmp17",
          "op": "const",
          "type": "bool",
          "value": true
        },
        {
          "args": [
            "tmp17"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "is_decreasing",
      "type": "bool"
    },
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "tmp18",
          "op": "const",
          "type": "int",
          "value": 10
        },
        {
          "args": [
            "x",
            "tmp18"
          ],
          "dest": "tmp19",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "dest": "tmp20",
          "op": "const",
          "type": "int",
          "value": 10
        },
        {
          "args": [
            "tmp19",
            "tmp20"
          ],
          "dest": "tmp21",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "x",
            "tmp21"
          ],
          "dest": "tmp22",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "tmp22"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "last_digit",
      "type": "int"
    }
  ]
}
