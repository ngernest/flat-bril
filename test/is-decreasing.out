@main
	op: call 		dest: tmp0		type: bool		args: ["x"]		funcs: ["is_decreasing"]
	op: id   		dest: tmp		type: bool		args: ["tmp0"]	
	op: print		args: ["tmp"]	
@is_decreasing
	op: id   		dest: tmp		type: int		args: ["x"]	
	op: const		dest: tmp1		type: int		value: 1	
	op: const		dest: tmp2		type: int		value: -1	
	op: mul  		dest: tmp3		type: int		args: ["tmp1", "tmp2"]	
	op: id   		dest: prev		type: int		args: ["tmp3"]	
.label4
	op: const		dest: tmp7		type: int		value: 0	
	op: gt   		dest: tmp8		type: bool		args: ["tmp", "tmp7"]	
	op: br   		args: ["tmp8"]		labels: ["label5", "label6"]
.label5
	op: call 		dest: tmp9		type: int		args: ["tmp"]		funcs: ["last_digit"]
	op: id   		dest: digit		type: int		args: ["tmp9"]	
	op: lt   		dest: tmp10		type: bool		args: ["digit", "prev"]	
	op: br   		args: ["tmp10"]		labels: ["label11", "label12"]
.label11
	op: const		dest: tmp14		type: bool		value: false	
	op: ret  		args: ["tmp14"]	
	op: jmp  		labels: ["label13"]
.label12
	op: jmp  		labels: ["label13"]
.label13
	op: id   		dest: prev		type: int		args: ["digit"]	
	op: const		dest: tmp15		type: int		value: 10	
	op: div  		dest: tmp16		type: int		args: ["tmp", "tmp15"]	
	op: id   		dest: tmp		type: int		args: ["tmp16"]	
	op: jmp  		labels: ["label4"]
.label6
	op: const		dest: tmp17		type: bool		value: true	
	op: ret  		args: ["tmp17"]	
@last_digit
	op: const		dest: tmp18		type: int		value: 10	
	op: div  		dest: tmp19		type: int		args: ["x", "tmp18"]	
	op: const		dest: tmp20		type: int		value: 10	
	op: mul  		dest: tmp21		type: int		args: ["tmp19", "tmp20"]	
	op: sub  		dest: tmp22		type: int		args: ["x", "tmp21"]	
	op: ret  		args: ["tmp22"]	
